### PR TITLE
WIP - Integer tensor

### PR DESCRIPTION
**Pull request by acbellini**
Work in progress for Integer tensor. 

- pulled up some methods from Integer Tensor to BaseTensor
- reduced the dependency of SyftController from FloatTensor. Unfortunately, being int and float not objects, it's not easy to actualy treat them as just different instances of the same class. We could use a generic T where T:struct, but it's really messy. As I don't expect we'll have many tensor types except Integer and Float, I think that the duplication of the methods is not too much trouble. 
- created IntegerTensor. Lots is still unimplemented and completely untested. Still missing the frontend
- fixed a bug in to_numpy in FloatTensor that would make it fail if the data is on the GPU. 
- modified SyftController so that it accepts a "integertensor" object type
- some placeholder code highlighted with FIXME (especially parsing the data from the client into an int[], at the moment it's a placeholder value)
- still to do: 
  - complete the implementation of IntegerTensor to the point of at least creating integer tensors on the CPU (no ops yet). 
  - do the python frontend

### CODE REVIEW

**Comment by acbellini** [comment]
The conflict on FloatTensorOps.cs should be resolved by deleting the method as it has been pulled up in BaseTensor

File: UnityProject/Assets/OpenMined/Syft/Tensor/BaseTensor.cs [hunk 1]
```diff
@@ -170,5 +171,65 @@ public int[] GetIndices(int index)
...
+
+        public void ZeroGPU_()
+        {
+            shader.SetBuffer(ZeroKernel_, "ZeroData_", dataBuffer);
+            shader.Dispatch(ZeroKernel_, this.size, 1, 1);
+        }
+
+
+
+        public bool IsContiguous()
```
Comment (justin1121) [inline UnityProject/Assets/OpenMined/Syft/Tensor/BaseTensor.cs]: There's a new version of IsContiguous in master that this should be

**Comment by justin1121** [comment]
> As I don't expect we'll have many tensor types except Integer and Float, I think that the duplication of the methods is not too much trouble.

The plan right now is to implement all the tensor types listed here http://pytorch.org/docs/master/tensors.html. So keeping things as generic as possible is probably the best bet I'd think.

**Comment by acbellini** [comment]
I'll take a look at the pytorch implementation. There is some issue with the fact that in C# int and float don't really share a common base type, only `struct` can be used, but then you cannot write something like

`T val where T:struct;`

meaning you can't restrict the type one a variable. I tried the fully generics way, but it ended up needing explicit casts everywhere. However, this may be due to my lack of C# skills :) I'll look to see if I can find a way to make things work better, I'm no fan of redundant code either :) Will report back!

File: UnityProject/Assets/OpenMined/Syft/Tensor/BaseTensor.cs (hunk 1 above)
Comment (acbellini) [reply UnityProject/Assets/OpenMined/Syft/Tensor/BaseTensor.cs]: no, the conflict is in FloatTensor.Ops.cs. Nothing problematic, I will resolve it once the PR is finalised.

**Comment by iamtrask** [comment]
Great work Anna! This is really coming together. We've had this topic of abstract types come up several times before, and we haven't come across a great option yet. It gets even more difficult at the GPU level. We'll almost definitely need to write custom GPU code for each type. All that is to say, given what we know about C# so far, I support your decisions listed above :)

File: UnityProject/Assets/OpenMined/Syft/Tensor/BaseTensor.cs (hunk 1 above)
Comment (justin1121) [reply UnityProject/Assets/OpenMined/Syft/Tensor/BaseTensor.cs]: Guess I meant where this function is in FloatTensor.Ops.cs. Whoops!

**Pull request merged**